### PR TITLE
Add jobs to the DB faster

### PR DESCRIPTION
**Pull request by seanh**
Add jobs to the DB faster

When adding `"sync_annotation"` jobs for all annotations between two dates to the DB, instead of using SQLAlchemy's ORM use its SQL expression language to generate a single `insert into job ... select ... from annotation where ...` SQL expression. This is much faster than using the ORM so it should enable larger numbers of jobs to be added to the DB without the admin page form submission request timing out.

This means that the annotation ID's stored in the `Job.kwargs` column are now in the DB's internal hex format instead of the application-level URL-safe format, so conversion to URL-safe format had to be added when reading these IDs from the `kwargs` column.

`add_annotations_between_times()` uses a hard-coded priority of `1000` because it's not possible to efficiently count the number of annotations being added for use as the priority. Now that we're using a single `insert into job ... select ... from annotation where ...` to both select the matching annotations and insert the jobs in a single statement, we no longer materialize a list of all the matching annotations at any time so we can't count them. To get a count we'd have to do a separate SQL statement. Just use a hard-coded priority instead, it's much simpler and faster.

### CODE REVIEW

File: h/services/search_index/_queue.py [hunk 1]
```diff
@@ -23,30 +25,31 @@ def __init__(self, db, es, batch_indexer):
...
+        :type start_time: datetime.datetime
+        :type end_time: datetime.datetime
+        """
+        self._db.execute(
+            Job.__table__.insert().from_select(
+                [Job.name, Job.priority, Job.tag, Job.kwargs],
+                select(
+                    [
+                        text("'sync_annotation'"),
+                        text("1000"),
```
Comment (seanh) [inline h/services/search_index/_queue.py:45]: No enum yet, just a hard-coded priority. I figure we can add an enum later when we have multiple priorities

File: h/services/search_index/_queue.py [hunk 2]
```diff
@@ -71,6 +74,14 @@ def sync(self, limit):
         if not jobs:
             return
 
+        # Annotation IDs are stored in the DB's internal hex format in the job
+        # table but we need them in the URL-safe format used by
+        # models.Annotation and in Elasticsearch.
+        for job in jobs:
+            job.kwargs["annotation_id"] = URLSafeUUID().process_result_value(
+                job.kwargs["annotation_id"], "dialect"
+            )
```
Comment (seanh) [inline h/services/search_index/_queue.py:83]: This modifies `job.kwargs["annotation_id"]` in Python but the modification won't be saved back to the DB because of the lack of mutation tracking for JSON columns. In this case that's what we actually want.

`"dialect"` is unused.

Might want to refactor the code in `types.py` so I don't have to do `URLSafeUUID().process_result_value(value, "dialect")`. Provide a `to_urlsafe(value)` function

File: h/services/search_index/_queue.py [hunk 3]
```diff
@@ -23,30 +25,31 @@ def __init__(self, db, es, batch_indexer):
...
+                select(
+                    [
+                        text("'sync_annotation'"),
+                        text("1000"),
+                        text(repr(tag)),
+                        func.jsonb_build_object("annotation_id", Annotation.id),
+                    ]
+                )
+                .where(Annotation.updated >= start_time)
+                .where(Annotation.updated <= end_time),
```
Comment (seanh) [inline h/services/search_index/_queue.py:51]: I think this is probably more or less as efficient as we're gonna get, in terms of adding annotations to the jobs table. So hopefully this is fast enough to reindex large timeframes

File: h/services/search_index/_queue.py [hunk 4]
```diff
@@ -44,11 +46,28 @@ def add_all(self, annotation_ids, tag, schedule_in=None):
                 name="sync_annotation",
                 scheduled_at=scheduled_at,
                 priority=priority,
-                kwargs={"annotation_id": annotation_id},
+                kwargs={"annotation_id": URLSafeUUID.url_safe_to_hex(annotation_id)},
             )
             for annotation_id in annotation_ids
         )
 
+    def add_annotations_between_times(self, start_time, end_time, tag):
```
Comment (seanh) [inline h/services/search_index/_queue.py:54]: The docstring for this method is on the corresponding `SearchIndexService.add_annotations_between_times()` since that's public and this is private

File: h/services/search_index/_queue.py [hunk 5]
```diff
@@ -44,11 +46,28 @@ def add_all(self, annotation_ids, tag, schedule_in=None):
                 name="sync_annotation",
                 scheduled_at=scheduled_at,
                 priority=priority,
-                kwargs={"annotation_id": annotation_id},
+                kwargs={"annotation_id": URLSafeUUID.url_safe_to_hex(annotation_id)},
```
Comment (seanh) [inline h/services/search_index/_queue.py:49]: The existing `add()` and `add_all()` methods are changed to use the hex IDs like `add_annotations_between_times()` does so that they're all consistent

File: h/services/search_index/_queue.py (hunk 3 above)
Comment (jon-betts) [reply h/services/search_index/_queue.py:68]: If it's not quick enough, we can always spin it off into an async task. Nicer if we don't have to bother

Comment (jon-betts) [reply h/services/search_index/_queue.py:68]: If it's not quick enough, we can always spin it off into an async task. Nicer if we don't have to bother

File: h/services/search_index/_queue.py [hunk 6]
```diff
@@ -71,7 +92,9 @@ def sync(self, limit):
         if not jobs:
             return
 
-        annotation_ids = {job.kwargs["annotation_id"] for job in jobs}
+        annotation_ids = {
+            URLSafeUUID.hex_to_url_safe(job.kwargs["annotation_id"]) for job in jobs
```
Comment (jon-betts) [inline h/services/search_index/_queue.py:96]: Maybe a comment here about the translation?

File: h/services/search_index/service.py [hunk 7]
```diff
@@ -77,15 +76,7 @@ def add_annotations_between_times(self, start_time, end_time, tag):
...
-        self._queue.add_all(
-            [
-                row[0]
-                for row in self._db.query(Annotation.id)
-                .filter(Annotation.updated >= start_time)
-                .filter(Annotation.updated <= end_time)
-            ],
-            tag=tag,
-        )
+        self._queue.add_annotations_between_times(start_time, end_time, tag)
```
Comment (jon-betts) [inline h/services/search_index/service.py:79]: Yep. No need to get fancy. Until we hit the rule of three probably not worth generalising here.

File: tests/h/services/search_index/_queue_test.py [hunk 8]
```diff
@@ -32,12 +35,54 @@ def test_it(self, db_session, queue, now):
...
                 )
             ),
         ]
 
 
+class TestAddAnnotationsBetweenTimes:
+    def test_it(self, annotation_ids, db_session, queue):
+        queue.add_annotations_between_times(
+            datetime_.datetime(2020, 9, 9),
+            datetime_.datetime(2020, 9, 11),
```
Comment (jon-betts) [inline tests/h/services/search_index/_queue_test.py:50]: Maybe use a different example date here?

**Review by jon-betts — APPROVED** [review]
This looks good to me, although we can't prove the main point without trying it with a lot of results .

There are a few mostly stylistic suggestions below. The only thing of substance I can think of is the tests for `add_annotations_between_times()` doesn't explore the value boundaries as closely as it should.

The examples intended not to be picked up are too far from the edge case to pick up.
You can see this in effect by changing the swapping the `>=, <=` for `>, <` and see the tests still pass.

It might be better to have tests with something like this:

```
START_DATE - timedelta(day=1)
START_DATE - timedelta(microsecond=1)
START_DATE
START_DATE + timedelta(microsecond=1)
```

To really dial in the behavior we want
